### PR TITLE
DEVEXP-291: Update callbacks for all APIs

### PR DESCRIPTION
**Pull request by asein-sinch**
- Add signature verification for Numbers API (algorithm found [here](https://gitlab.com/sinch/sinch-projects/enterprise-and-messaging/sms/applications/selfserve/numbers/numbers-callback-service/-/blob/dev/callback-worker/src/main/java/com/sinch/numbers/callback/worker/CallbackSender.java#L137) and [here](https://gitlab.com/sinch/sinch-projects/enterprise-and-messaging/sms/applications/selfserve/numbers/numbers-callback-service/-/blob/dev/callback-worker/src/main/java/com/sinch/numbers/callback/util/EncryptionUtils.java#L9))
 - Homogenise callback handling across the various API
 - Add documentation for the Nest.js app

### CODE REVIEW

File: examples/simple-examples/src/numbers/available/rentAny.ts [hunk 1]
```diff
@@ -17,6 +18,7 @@ import {
 
   const servicePlanId = readServicePlanId();
   const appId = readApplicationKey();
+  const callbackUrl = getNumberCallbackUrlFromConfig();
 
   if (!servicePlanId && !appId) {
```
Comment (JPPortier) [inline examples/simple-examples/src/numbers/available/rentAny.ts:23]: Does this test is not managed by client itself ?

File: examples/webhooks/src/controller/app.controller.ts [hunk 2]
```diff
@@ -26,10 +26,20 @@ export class AppController {
     private readonly voiceService: VoiceService) {}
 
   @Post('/numbers')
-  public numbers(@Body() eventBody: any, @Res() res: Response) {
-    // console.log(eventBody);
+  public numbers(@Req() request: Request, @Res() res: Response) {
+    // Initialize the class that will be used to validate the request and parse it
+    const numbersCallbackWebhook = new NumbersCallbackWebhooks(SINCH_NUMBERS_CALLBACK_SECRET);
+    // 1 - The first thing to do is to verify the request is legit and has not been tampered
+    const validated = numbersCallbackWebhook.validateRequestIntegrity(
```
Comment (JPPortier) [inline examples/webhooks/src/controller/app.controller.ts:33]: `validateAuthenticationHeader` ?

File: examples/webhooks/src/controller/app.controller.ts [hunk 3]
```diff
@@ -26,10 +26,20 @@ export class AppController {
...
+      request.headers, request['rawBody']);
+    if (!validated) {
+      res.status(401).send('Invalid signature');
+      return;
+    }
     try {
-      const event = parseNumbersEventNotification(eventBody);
+      // 2 - Before acting on the request, it must be parsed to verify it's supported and to revive its content
+      const event = numbersCallbackWebhook.parseNumbersEventNotification(request.body);
```
Comment (JPPortier) [inline examples/webhooks/src/controller/app.controller.ts:41]: `parseEvent(warning)` ?

File: examples/webhooks/src/controller/app.controller.ts [hunk 4]
```diff
@@ -40,9 +50,11 @@ export class AppController {
 
   @Post('/sms')
   public sms(@Body() eventBody: any, @Res() res: Response) {
-    // console.log(eventBody);
+    const smsCallbackWebhooks = new SmsCallbackWebhooks();
     try {
-      const event = parseSmsEventNotification(eventBody);
+      // There is no request validation for the SMS API, so we can parse it and revive its content directly
+      const event = smsCallbackWebhooks.parseSmsEventNotification(eventBody);
```
Comment (JPPortier) [inline examples/webhooks/src/controller/app.controller.ts:56]: `parseEvent(warning)` ?

File: examples/webhooks/src/controller/app.controller.ts [hunk 5]
```diff
@@ -53,21 +65,22 @@ export class AppController {
...
-    // console.log(request['rawBody']);
-    const sinchClient = new SinchClient({
+    // Initialize the class that will be used to validate the request and parse it
+    const verificationCallbackWebhooks = new VerificationCallbackWebhooks({
       applicationKey: SINCH_APPLICATION_KEY,
       applicationSecret: SINCH_APPLICATION_SECRET
     });
-    const validated = sinchClient.verification.callbacks.validateAuthorizationHeader(
+    // 1 - The first thing to do is to verify the request is legit and has not been tampered
+    const validated = verificationCallbackWebhooks.validateAuthorizationHeader(
```
Comment (JPPortier) [inline examples/webhooks/src/controller/app.controller.ts:74]: `validateAuthenticationHeader` ?

File: examples/webhooks/src/controller/app.controller.ts [hunk 6]
```diff
@@ -77,18 +90,22 @@ export class AppController {
...
-    // console.log(request.body);
-    // console.log(request['rawBody']);
-    const validated = validateAuthenticationHeader(SINCH_APPLICATION_KEY, SINCH_APPLICATION_SECRET,
+    // Initialize the class that will be used to validate the request and parse it
+    const voiceCallbackWebhooks = new VoiceCallbackWebhooks({
+      applicationKey: SINCH_APPLICATION_KEY,
+      applicationSecret: SINCH_APPLICATION_SECRET
+    });
+    // 1 - The first thing to do is to verify the request is legit and has not been tampered
+    const validated = voiceCallbackWebhooks.validateAuthorizationHeader(
```
Comment (JPPortier) [inline examples/webhooks/src/controller/app.controller.ts:99]: `validateAuthenticationHeader` ?

File: examples/simple-examples/src/numbers/available/rentAny.ts (hunk 1 above)
Comment (asein-sinch) [reply examples/simple-examples/src/numbers/available/rentAny.ts:23]: No, here the servicePlanId and the appId values are read from the configuration to assign a number to a SMS app or a Voice app.
As we are in the number app, the sinch client is using the projectId credentials and doesn't know about the SMS app(s) and Voice app(s)

Comment (JPPortier) [reply examples/simple-examples/src/numbers/available/rentAny.ts:23]: Ok, thank you, I missed this point

File: examples/webhooks/src/controller/app.controller.ts (hunk 2 above)
Comment (asein-sinch) [reply examples/webhooks/src/controller/app.controller.ts]: Fixed

File: examples/webhooks/src/controller/app.controller.ts (hunk 3 above)
Comment (asein-sinch) [reply examples/webhooks/src/controller/app.controller.ts]: Fixed

File: examples/webhooks/src/controller/app.controller.ts (hunk 4 above)
Comment (asein-sinch) [reply examples/webhooks/src/controller/app.controller.ts]: Fixed

File: examples/webhooks/src/controller/app.controller.ts (hunk 6 above)
Comment (asein-sinch) [reply examples/webhooks/src/controller/app.controller.ts]: Fixed

**Pull request merged**